### PR TITLE
feat: add dapp-portal support to zk_inception

### PR DESCRIPTION
**Pull request by sanekmelnikov**
## What❔

This PR introduces a new `portal` subcommand to the `zk_inception` CLI tool, enabling users to easily launch the [dapp-portal](https://github.com/matter-labs/dapp-portal) for their deployed chains.

Usage: `zk_inception portal [--port 3030]`

The ecosystem configurations are automatically converted to the [hyperchains](https://github.com/matter-labs/dapp-portal/tree/main/hyperchains#%EF%B8%8F-configure-manually) format, which is used to configure dapp-portal at runtime. Essentially, the following command is executed under the hood:
`docker run -p PORT:3000 /path/to/portal.config.js:/usr/src/app/dist/config.js dapp-portal`

## Why ❔

Currently, running the dapp-portal requires users to manually pull the repository, install all dependencies, modify configurations, build the project, and then run it - a tedious and time-consuming process. This PR simplifies the process, allowing users to run the portal effortlessly with a single command.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct TokenConfig {
+    pub address: String,
+    pub symbol: String,
+    pub decimals: u8,
+    #[serde(rename = "l1Address", skip_serializing_if = "Option::is_none")]
+    pub l1_address: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+}
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/portal.rs:90]: Please move it to the config crate

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+                public: RpcUrlConfig {
+                    http: vec![l1_rpc_url.to_string()],
+                },
+            },
+        })
+    } else {
+        None
+    };
+    // TODO: use bridge contract to derive base token
+    // TODO: add ERC20 tokens
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/portal.rs:135]: base token on l2 always has an address `0x8002`

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    // Tokens: add ETH token
+    let tokens = vec![
+        TokenConfig {
+            address: L2_BASE_TOKEN_ADDRESS.to_string(),
+            l1_address: Some("0x0000000000000000000000000000000000000000".to_string()),
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/portal.rs:140]: it's not true for chains with custom base token.

Comment (sanekmelnikov) [reply zk_toolbox/crates/zk_inception/src/commands/portal.rs]: now getting it from config
interesting edge-case - what happens when L1 base token has e.g. 6 decimals, I assume L2 base token would still have 18? (e.g. some [contracts](https://github.com/matter-labs/era-contracts/blob/8670004d6daa7e8c299087d62f1451a3dec4f899/l2-contracts/contracts/bridge/L2WrappedBaseToken.sol#L43) rely on?)

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs (hunk 2 above)
Comment (sanekmelnikov) [reply zk_toolbox/crates/zk_inception/src/commands/portal.rs]: yes, added 0x800A value to L2_BASE_TOKEN_ADDRESS as a constant

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs (hunk 1 above)
Comment (sanekmelnikov) [reply zk_toolbox/crates/zk_inception/src/commands/portal.rs]: moved

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                },
+            },
+        })
+    } else {
+        None
+    };
+    // Base token:
+    let (base_token_addr, base_token_info) = if chain_config.base_token == BaseToken::eth() {
+        (
+            ethers::utils::hex::encode_prefixed(Address::zero()),
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/portal.rs:68]: i thing, that `format!("{Address::zero():?}")` will have the exact same result

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        // Initialize non-public L1 network
+        Some(L1NetworkConfig {
+            id: chain_config.l1_network.chain_id(),
+            name: chain_config.l1_network.to_string(),
+            network: chain_config.l1_network.to_string().to_lowercase(),
+            native_currency: NativeCurrency {
+                name: "Ether".to_string(),
+                symbol: "ETH".to_string(),
+                decimals: 18,
+            },
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/portal.rs:52]: I'd prefer to have it as a method for native currency. and what is the difference between `TokenInfo` and `NativeCurrency`?

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    }
+    if chain_configs.is_empty() {
+        return Err(anyhow!(MSG_FAILED_TO_START_PORTAL_ERR));
+    }
+
+    let hyperchain_configs = create_hyperchains_config(&chain_configs).await?;
+    let runtime_config = PortalRuntimeConfig {
+        node_type: "hyperchain".to_string(),
+        hyperchain_configs,
+    };
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/portal.rs:146]: should we move it to the `init` process and regenerate it only when we add a new chain?

Comment (sanekmelnikov) [reply zk_toolbox/crates/zk_inception/src/commands/portal.rs]: moved

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs (hunk 5 above)
Comment (sanekmelnikov) [reply zk_toolbox/crates/zk_inception/src/commands/portal.rs]: replaced NativeCurrency with TokenInfo

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs (hunk 4 above)
Comment (sanekmelnikov) [reply zk_toolbox/crates/zk_inception/src/commands/portal.rs]: replaced with `format`

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        .api_config
+        .as_ref()
+        .context("api_config")?
+        .web3_json_rpc
+        .http_url;
+    // Check if L1 network is public
+    let public_l1_network_id = match chain_config.l1_network {
+        L1Network::Sepolia | L1Network::Mainnet => Some(chain_config.l1_network.chain_id()),
+        _ => None,
+    };
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/portal.rs:36]: Why do we need to distinguish it here?

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    let secrets_config = chain_config.get_secrets_config()?;
+    let l1_rpc_url = secrets_config
+        .l1
+        .as_ref()
+        .context("l1")?
+        .l1_rpc_url
+        .expose_str();
+    let l1_network = if public_l1_network_id.is_none() {
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/portal.rs:44]: you could just use `map` function for it

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    let chains: Vec<String> = ecosystem_config.list_of_chains();
+    let mut chain_configs = Vec::new();
+    for chain in chains {
+        if let Some(chain_config) = ecosystem_config.load_chain(Some(chain.clone())) {
+            chain_configs.push(chain_config)
+        }
+    }
+    let hyperchains_config = create_hyperchains_config(&chain_configs).await?;
+    if hyperchains_config.is_empty() {
+        return Err(anyhow!("Failed to create any valid hyperchain config"));
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/portal.rs:120]:
```suggestion
       anyhow::bail!("Failed to create any valid hyperchain config"))
```

File: zk_toolbox/crates/zk_inception/src/commands/chain/init.rs [hunk 10]
```diff
@@ -126,6 +129,8 @@ pub async fn init(
         .await
         .context(MSG_GENESIS_DATABASE_ERR)?;
 
+    let _ = create_and_save_portal_config(ecosystem_config, shell).await;
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/chain/init.rs:132]: why ignoring?

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs (hunk 7 above)
Comment (sanekmelnikov) [reply zk_toolbox/crates/zk_inception/src/commands/portal.rs:36]: it allows to use default config for it (like prefilled block explorer addresses, etc.) [[link](https://github.com/matter-labs/dapp-portal/blob/485d8791fd7003f1bf7b65b9c7c6515c95511ebd/hyperchains/README.md?plain=1#L49)]

File: zk_toolbox/crates/zk_inception/src/commands/chain/init.rs (hunk 10 above)
Comment (Deniallugo) [reply zk_toolbox/crates/zk_inception/src/commands/chain/init.rs:132]: why could it fail though? 
i thought we are only creating configs there

File: zk_toolbox/crates/zk_inception/src/commands/portal.rs (hunk 7 above)
Comment (Deniallugo) [reply zk_toolbox/crates/zk_inception/src/commands/portal.rs:36]: I don't think it brings any simplicity in our case.

File: zk_toolbox/crates/zk_inception/src/consts.rs [hunk 11]
```diff
@@ -7,3 +7,6 @@ pub const PROVER_STORE_MAX_RETRIES: u16 = 10;
 pub const DEFAULT_CREDENTIALS_FILE: &str = "~/.config/gcloud/application_default_credentials.json";
 pub const DEFAULT_PROOF_STORE_DIR: &str = "artifacts";
 pub const BELLMAN_CUDA_DIR: &str = "era-bellman-cuda";
+pub const L2_BASE_TOKEN_ADDRESS: &str = "0x000000000000000000000000000000000000800A";
+pub const PORTAL_DOCKER_IMAGE: &str = "matterlabs/dapp-portal";
+pub const PORTAL_DOCKER_PORT: u16 = 3000;
```
[the remark this answers is not in the document]
Comment (sanekmelnikov) [reply zk_toolbox/crates/zk_inception/src/consts.rs]: that's docker container (internal) port, i'll change the const name for clarity

**Review by Deniallugo — APPROVED** [review]

**Pull request merged**